[JDBC persistence MariaDB] binding causes errors at "restoreOnStartup" restore for Location and Image Items

Hi there,
after upgrading from OH v2.5.12 directly to OH v3.1.0-M3 the JDBC persistence MariaDB binding causes errors in karaf.log when restoring Image and Location Items at startup. The following errors only occur once per item during startup phase. I have two examples for each:

Location Items:
```
2021-04-08 13:21:28.442 [DEBUG] [jdbc.internal.JdbcPersistenceService] - JDBC::query: item is locationJR
2021-04-08 13:21:28.443 [DEBUG] [persistence.jdbc.internal.JdbcMapper] - JDBC::getHistItemFilterQuery filter='true' numberDecimalcount='3' table='Item372' item='locationJR (Type=LocationItem, State=NULL, Label=JR Location, Category=null)' itemName='locationJR'
2021-04-08 13:21:28.444 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::getHistItemFilterQueryProvider filter = org.openhab.core.persistence.FilterCriteria@17ac9388, numberDecimalcount = 3, table = Item372, simpleName = locationJR
2021-04-08 13:21:28.444 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::query queryString = SELECT time, value FROM Item372 ORDER BY time DESC  LIMIT 0,1
2021-04-08 13:21:28.445 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::doGetHistItemFilterQuery sql=SELECT time, value FROM Item372 ORDER BY time DESC  LIMIT 0,1
2021-04-08 13:21:28.448 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::ItemResultHandler::handleResult getState value = '49.9725155,8.3393236', unit = 'null', getClass = 'class java.lang.String', clazz = 'String'
2021-04-08 13:21:28.449 [DEBUG] [jdbc.internal.JdbcPersistenceService] - JDBC::query: query for locationJR returned 1 rows in 6 ms
2021-04-08 13:21:28.450 [ERROR] [org.openhab.core.items.GenericItem  ] - Tried to set invalid state 51.5085,-0.1257 (StringType) on item locationJR of type LocationItem, ignoring it
…
2021-04-08 13:21:28.491 [DEBUG] [jdbc.internal.JdbcPersistenceService] - JDBC::query: item is localStationLocation
2021-04-08 13:21:28.492 [DEBUG] [persistence.jdbc.internal.JdbcMapper] - JDBC::getHistItemFilterQuery filter='true' numberDecimalcount='3' table='Item40' item='localStationLocation (Type=LocationItem, State=NULL, Label=Location, Category=location)' itemName='localStationLocation'
2021-04-08 13:21:28.493 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::getHistItemFilterQueryProvider filter = org.openhab.core.persistence.FilterCriteria@7ac8c3b8, numberDecimalcount = 3, table = Item40, simpleName = localStationLocation
2021-04-08 13:21:28.494 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::query queryString = SELECT time, value FROM Item40 ORDER BY time DESC  LIMIT 0,1
2021-04-08 13:21:28.495 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::doGetHistItemFilterQuery sql=SELECT time, value FROM Item40 ORDER BY time DESC  LIMIT 0,1
2021-04-08 13:21:28.497 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::ItemResultHandler::handleResult getState value = '49.9725,8.3393', unit = 'null', getClass = 'class java.lang.String', clazz = 'String'
2021-04-08 13:21:28.498 [DEBUG] [jdbc.internal.JdbcPersistenceService] - JDBC::query: query for localStationLocation returned 1 rows in 6 ms
2021-04-08 13:21:28.499 [ERROR] [org.openhab.core.items.GenericItem  ] - Tried to set invalid state 51.5085,-0.1257 (StringType) on item localStationLocation of type LocationItem, ignoring it
```

Image Items:

```
2021-04-08 13:21:28.292 [DEBUG] [jdbc.internal.JdbcPersistenceService] - JDBC::query: item is MyCityCurrentConditionIcon
2021-04-08 13:21:28.292 [DEBUG] [persistence.jdbc.internal.JdbcMapper] - JDBC::getHistItemFilterQuery filter='true' numberDecimalcount='3' table='Item25' item='MyCityCurrentConditionIcon (Type=ImageItem, State=NULL, Label=Icon, Category=null)' itemName='MyCityCurrentConditionIcon'
2021-04-08 13:21:28.293 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::getHistItemFilterQueryProvider filter = org.openhab.core.persistence.FilterCriteria@1b63a0fa, numberDecimalcount = 3, table = Item25, simpleName = MyCityCurrentConditionIcon
2021-04-08 13:21:28.294 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::query queryString = SELECT time, value FROM Item25 ORDER BY time DESC  LIMIT 0,1
2021-04-08 13:21:28.295 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::doGetHistItemFilterQuery sql=SELECT time, value FROM Item25 ORDER BY time DESC  LIMIT 0,1
2021-04-08 13:21:28.298 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::ItemResultHandler::handleResult getState value = '[114, 97, 119, 32, 116, 121, 112, 101, 32, 40, 105, 109, 97, 103, 101, 47, 112, 110, 103, 41, 58, 32, 50, 55, 55, 51, 32, 98, 121, 116, 101, 115]', unit = 'null', getClass = 'class [B', clazz = 'byte[]'
2021-04-08 13:21:28.299 [ERROR] [nal.common.AbstractInvocationHandler] - An error occurred while calling method 'QueryablePersistenceService.query()' on 'org.openhab.persistence.jdbc.internal.JdbcPersistenceService@6bfdcace': class [B cannot be cast to class java.lang.String ([B and java.lang.String are in module java.base of loader 'bootstrap')
2021-04-08 13:21:28.317 [DEBUG] [jdbc.internal.JdbcPersistenceService] - JDBC::getName: returning name 'jdbc' for queryable persistence service.
2021-04-08 13:21:28.317 [ERROR] [ence.internal.PersistenceManagerImpl] - Exception occurred while querying persistence service 'jdbc': class [B cannot be cast to class java.lang.String ([B and java.lang.String are in module java.base of loader 'bootstrap')
…
2021-04-08 13:21:28.583 [DEBUG] [jdbc.internal.JdbcPersistenceService] - JDBC::query: item is MyCityHourlyForecast03ConditionIcon
2021-04-08 13:21:28.584 [DEBUG] [persistence.jdbc.internal.JdbcMapper] - JDBC::getHistItemFilterQuery filter='true' numberDecimalcount='3' table='Item45' item='MyCityHourlyForecast03ConditionIcon (Type=ImageItem, State=NULL, Label=Icon, Category=null)' itemName='MyCityHourlyForecast03ConditionIcon'
2021-04-08 13:21:28.584 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::getHistItemFilterQueryProvider filter = org.openhab.core.persistence.FilterCriteria@515081eb, numberDecimalcount = 3, table = Item45, simpleName = MyCityHourlyForecast03ConditionIcon
2021-04-08 13:21:28.585 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::query queryString = SELECT time, value FROM Item45 ORDER BY time DESC  LIMIT 0,1
2021-04-08 13:21:28.586 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::doGetHistItemFilterQuery sql=SELECT time, value FROM Item45 ORDER BY time DESC  LIMIT 0,1
2021-04-08 13:21:28.588 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::ItemResultHandler::handleResult getState value = '[114, 97, 119, 32, 116, 121, 112, 101, 32, 40, 105, 109, 97, 103, 101, 47, 112, 110, 103, 41, 58, 32, 50, 55, 55, 51, 32, 98, 121, 116, 101, 115]', unit = 'null', getClass = 'class [B', clazz = 'byte[]'
2021-04-08 13:21:28.589 [ERROR] [nal.common.AbstractInvocationHandler] - An error occurred while calling method 'QueryablePersistenceService.query()' on 'org.openhab.persistence.jdbc.internal.JdbcPersistenceService@6bfdcace': class [B cannot be cast to class java.lang.String ([B and java.lang.String are in module java.base of loader 'bootstrap')
2021-04-08 13:21:28.594 [DEBUG] [jdbc.internal.JdbcPersistenceService] - JDBC::getName: returning name 'jdbc' for queryable persistence service.
2021-04-08 13:21:28.595 [ERROR] [ence.internal.PersistenceManagerImpl] - Exception occurred while querying persistence service 'jdbc': class [B cannot be cast to class java.lang.String ([B and java.lang.String are in module java.base of loader 'bootstrap')
```

I circumvented the issue by removing the "restoreOnStartup" option in the jdbc.persist and activated MapDB for this purpose.
I saw that this problem has already appeared in the past: https://github.com/openhab/openhab1-addons/issues/5761 and maybe it has slipped back into OH with the upgrade to V3.

I have investigated this issue in the openHAB community under the topic: [https://community.openhab.org/t/oh-3-1-m3-persistence-error-for-location-items-is-back-jdbc-restoreonstartup-throw-error-for-location-items/120430](url) and have been advised to open an issue here.

Thank you for considering a solution!
Cheers ohb1138

--- a/bundles/org.openhab.persistence.jdbc/src/main/java/org/openhab/persistence/jdbc/db/JdbcBaseDAO.java
+++ b/bundles/org.openhab.persistence.jdbc/src/main/java/org/openhab/persistence/jdbc/db/JdbcBaseDAO.java
@@ -31,27 +31,25 @@ import org.eclipse.jdt.annotation.Nullable;
 import org.knowm.yank.Yank;
 import org.openhab.core.items.GroupItem;
 import org.openhab.core.items.Item;
-import org.openhab.core.library.items.ColorItem;
 import org.openhab.core.library.items.ContactItem;
 import org.openhab.core.library.items.DateTimeItem;
 import org.openhab.core.library.items.DimmerItem;
+import org.openhab.core.library.items.ImageItem;
 import org.openhab.core.library.items.NumberItem;
+import org.openhab.core.library.items.PlayerItem;
 import org.openhab.core.library.items.RollershutterItem;
-import org.openhab.core.library.items.StringItem;
 import org.openhab.core.library.items.SwitchItem;
 import org.openhab.core.library.types.DateTimeType;
 import org.openhab.core.library.types.DecimalType;
-import org.openhab.core.library.types.HSBType;
-import org.openhab.core.library.types.OnOffType;
-import org.openhab.core.library.types.OpenClosedType;
 import org.openhab.core.library.types.PercentType;
 import org.openhab.core.library.types.QuantityType;
-import org.openhab.core.library.types.StringType;
+import org.openhab.core.library.types.RawType;
 import org.openhab.core.library.unit.Units;
 import org.openhab.core.persistence.FilterCriteria;
 import org.openhab.core.persistence.FilterCriteria.Ordering;
 import org.openhab.core.persistence.HistoricItem;
 import org.openhab.core.types.State;
+import org.openhab.core.types.TypeParser;
 import org.openhab.persistence.jdbc.model.ItemVO;
 import org.openhab.persistence.jdbc.model.ItemsVO;
 import org.openhab.persistence.jdbc.model.JdbcHistoricItem;
@@ -505,23 +503,17 @@ public class JdbcBaseDAO {
             }
             return unit == null ? DecimalType.valueOf(((String) v).toString())
                     : QuantityType.valueOf(((String) v).toString());
-        } else if (item instanceof ColorItem) {
-            return HSBType.valueOf(((String) v).toString());
-        } else if (item instanceof DimmerItem) {
-            return new PercentType(objectAsInteger(v));
-        } else if (item instanceof SwitchItem) {
-            return OnOffType.valueOf(((String) v).toString().trim());
-        } else if (item instanceof ContactItem) {
-            return OpenClosedType.valueOf(((String) v).toString().trim());
-        } else if (item instanceof RollershutterItem) {
-            return new PercentType(objectAsInteger(v));
         } else if (item instanceof DateTimeItem) {
             return new DateTimeType(
                     ZonedDateTime.ofInstant(Instant.ofEpochMilli(objectAsLong(v)), ZoneId.systemDefault()));
-        } else if (item instanceof StringItem) {
-            return StringType.valueOf(((String) v).toString());
-        } else {// Call, Image, Location, Player, String
-            return StringType.valueOf(((String) v).toString());
+        } else if (item instanceof DimmerItem || item instanceof RollershutterItem) {
+            return new PercentType(objectAsInteger(v));
+        } else if (item instanceof ImageItem) {
+            return RawType.valueOf(objectAsString(v));
+        } else if (item instanceof ContactItem || item instanceof PlayerItem || item instanceof SwitchItem) {
+            return TypeParser.parseState(item.getAcceptedDataTypes(), ((String) v).toString().trim());
+        } else {
+            return TypeParser.parseState(item.getAcceptedDataTypes(), ((String) v).toString());
         }
     }
 
@@ -546,6 +538,13 @@ public class JdbcBaseDAO {
             return ((Byte) v).intValue();
         }
         return ((Integer) v).intValue();
+    }
+
+    protected String objectAsString(Object v) {
+        if (v instanceof byte[]) {
+            return new String((byte[]) v);
+        }
+        return ((String) v).toString();
     }
 
     public String getItemType(Item i) {

